# Aula 2  - Erros Propagados e método da bissecção 

### Fórmula Fundamental para o Cálculo do Erro

 Seja $\bar x$ um valor aproximado de $x$ com majorante do
  erro $\Delta x$ e uma função $f:[a,b]\subset\mathbb{R}\longrightarrow\mathbb{R}$
  regular ($C^1([a,b]$) no
  intervalo $[a,b]$ com $[\bar x-\Delta x,\bar x+\Delta x]\subset]a,b[$.
  
  Então um majorante
  do erro propagado por meio de $f$ é dado por
  $$\Delta y=\left|\frac{{\rm d}y}{{\rm d} x}\right|_M\Delta x,$$
onde $\left|\frac{{\rm d}y}{{\rm d} x}\right|_M$ é um majorante do
módulo  de $\displaystyle \frac{{\rm d}y}{{\rm d} x}$ em $[\bar x-\Delta x,\bar x+\Delta x]$.

### Exercício sobre o erro propagado

Seja $y=3x^2+1$ e $x= 2.7\pm 0.05$. Calcule $\bar y$ e $\Delta y$.


In [3]:
xa=2.7 # valor aproximado para xa está em [2.7-0.05, 2.7+0.05]
Dx=0.05 # Erro para x

def y(x): #Função que propaga o erro
    return 3*x**2+1

def dydx(x):
    # dydx é positiva no intervalo de interesse
    # f é crescente
    return 6*x

maj_dydx=6*(xa+Dx)

ya=y(xa)
Dy=maj_dydx*Dx

print('o valor aproximado de y é', ya)
print('o erro de y é', Dy)

o valor aproximado de y é 22.870000000000005
o erro de y é 0.8250000000000001


## Fórmula Fundamental do Cálculo do Erro. (FFCE)


Seja
  $\bar P=(\bar x_1,\dots,\bar x_n)\in\mathbb{R}^n\, (n\in\mathbb{N})$ um valor
  aproximado de $P$ com majorantes dos erros associados
  $\Delta x_i\, (i=1,\dots,n)$ e uma função $f:D\subset \mathbb{R}^n\longrightarrow\mathbb{R}$
  de classe $C^1(D)$.  Então um majorante do erro
  propagado por meio de $f$ é dado por
  $$\Delta y=\left|\frac{\partial y}{\partial x_1}\right|_M\Delta x_1+\cdots+\left|\frac{\partial y}{\partial x_n}\right|_M\Delta x_n,$$
  onde $\left|\frac{\partial y}{\partial x_i}\right|_M\, (i=1,\dots,n)$
  é um majorante do módulo de
  $\displaystyle \frac{\partial y}{\partial x_i}$ em
  $[\bar x_1-\Delta x_1,\bar x_1+\Delta x_1]\times \cdots\times[\bar
  x_n-\Delta x_n,\bar x_n+\Delta x_n]$.


### Exercício sobre o erro propagado.

 Seja $\large w=\frac{\pi r^2}{3\sqrt{a}}$, com $r=2.18\pm0.0015,$ $a=5.2\pm0.07$ e $\pi=3.14\pm0.001$.


 Calcule a aproximação de $w$ bem como majorantes para os erros propagados absoluto $\Delta w$ e relativo $\delta w$.

#### Resolução recorrendo à decomposição em sub-operações

Consideremos a primeira sub-operação definida por $x=r^2.$
O erro $\Delta x$ bem como $\bar x$ (aproximação de $x$)  são dados por:

In [4]:
r_=2.18 # valor aproximado para r
a_=5.2  # valor aproximado para a
pi_=3.14 # valor aproximado para pi
Dr=0.0015 # Erro absoluto para r
Da=0.07 # Erro absoluto para a
Dpi=0.001 # Erro absoluto para 
Dx=2*(r_+Dr)*Dr # majorante para erro da 1ª sub-operação
x_=r_**2 # Aproximação para o valor de x 
print('Dx=',Dx)
print('x_=',x_)

Dx= 0.006544500000000001
x_= 4.752400000000001


$$\Delta x= \left|3 r^2\right|_M\Delta r \leq 0.0066, \quad \bar
x=4.7524$$


Para segunda operação assumimos $y =\pi\times x$. Assim, o erro $\Delta y$ bem como $\bar y$ são dados por:

In [5]:
Dy=(pi_+Dpi)*Dx+(x_+Dx)*Dpi
y_=3.14*x_
print('Dy=',Dy)
print('y_=', y_)

Dy= 0.025315219000000003
y_= 14.922536000000003


$$\Delta y=|x|_M\Delta \pi +|\pi|_M\Delta x=0.0254,\quad \bar y=14.9225 $$


 A terceira sub-operação é definida por $z=3\sqrt{a}$.
O erro $\Delta z$ bem como $\bar z$ são dados por:


In [6]:
from math import sqrt
Dz=3.0/(2.*sqrt(a_-Da))*Da
z_=3.0*sqrt(a_)
print('Dz=',Dz)
print('z_=',z_)

Dz= 0.046358632497276536
z_= 6.841052550594828



$$\Delta z=4\left|\frac{1}{2\sqrt{a}}\right|_M\Delta
  a=0.0464,\quad \bar z=6.8411$$

A última sub-operação é definida por $w=\frac {y}{z}.$
O erro $\Delta w$ bem como $\bar w$ são dados por:

In [7]:
Dw=1./(z_-Dz)*Dy+(y_+Dy)/(z_-Dz)**2*Dz
w_=y_/z_
print('Dw=',Dw)
print('w_=',w_)

Dw= 0.018735349741747943
w_= 2.1813216445329733


Vamos arrendodar os valores de forma a presentar os resultados com 4 algarismos decimais

In [8]:
DeltaW=round(Dw+0.00005,4) #Arredondamento por excesso???? Porquê ????
print('DeltaW=',DeltaW)
W=round(w_,4) #Arredondamento Simétrico
print('w_=',W)

DeltaW= 0.0188
w_= 2.1813


$$\Delta w=\left|\frac{1}{z}\right|_M\Delta
  y+\left|\frac{y}{z^2}\right|_M\Delta
  z=0.0188,\quad \bar w=2.1813$$

Calculemos agora um majorante do erro relativo $\delta w=\frac{\Delta w}{|w|}$.
Note-se que:
- Conhecemos um majorante do erro absoluto;
- Não conhecemos o valor exacto de $w$;
- É necessário minorar o valor de $w$ para majorar o erro relativo (aqui em módulo). 

In [11]:
w_min=w_-Dw
print('w_min=',w_min)

delta_w=DeltaW/w_min
delta_w=round(delta_w+0.00005,4) # Outra vez por excesso
print('delta_w=',delta_w)

w_min= 3.5002083378091213
delta_w= 0.0093


Um majorante do erro relativo é dado por: $\delta w=\frac{\Delta w}{|w|}=0.0093$.

Em percentagem : $\delta w= 0.93 \%$.

## Resolução de equações não-lineares $f(x)=0$.
-  L. Burden, J. Douglas Faires, Numerical Analysis, 8th edition, Thomson, Brooks/Cole. **Cap. 2**

# O Método da Bissecção.

#### Exercício sobre o método da Bissecção:

- Prove que $\cos(x)=x$ tem uma única solução em $\mathbb{R}$ e, por aplicação do método da bissecção, apresente uma aproximação dessa solução com um erro inferior a $10^{-7}$

In [14]:
A=0.0
B=1.0

from math import  cos

#Função auxiliar
def f(x):
 return cos(x)-x

In [15]:
f(A)*f(B)

-0.45969769413186023

In [17]:
#Provou-se que a única solução está em [0,1]
#Com quantas iterações conseguimos garantir 

from math import log
#Extremos iniciais do intervalo
N=log(0.0000001/(B-A))/log(1./2.)
N=int(round(N+0.5,0))
print('N=',N) 



N= 24


In [18]:
#Ciclo para calcular as aproximações
for i in range(1,N+1):
 p=(A+B)/2.0
 if (f(p)==0) : # era sorte a mais, Não serve de grande coisa !
  break        # Se não tivessemos o número de iterações  a  
 else:         #paragem seria com o  erro (discutir Pseudo-Código)
  if (f(A)*f(p)>0):
   A=p
  else:
   B=p
 print('i=%d, p=%.7f' %(i, p))

i=1, p=0.5000000
i=2, p=0.7500000
i=3, p=0.6250000
i=4, p=0.6875000
i=5, p=0.7187500
i=6, p=0.7343750
i=7, p=0.7421875
i=8, p=0.7382812
i=9, p=0.7402344
i=10, p=0.7392578
i=11, p=0.7387695
i=12, p=0.7390137
i=13, p=0.7391357
i=14, p=0.7390747
i=15, p=0.7391052
i=16, p=0.7390900
i=17, p=0.7390823
i=18, p=0.7390862
i=19, p=0.7390842
i=20, p=0.7390852
i=21, p=0.7390847
i=22, p=0.7390850
i=23, p=0.7390851
i=24, p=0.7390851


### Como reaproveitar o código para outros casos?

- O método depende de:
 - função $f(x)$
 - Intervalo Inicial [A,B]
 - Condição de Paragem: erro e/ou número de iterações
 
 Vejamos um exemplo: criar um *módulo* : !Sair mode slideshow!

In [19]:
%reset
## Modulo bissect
## Condicao de paragem com erro=TOL ou Numero Maximo de iter. 
## Guardar sols num vector
import numpy as np

def bissect(f,A,B,TOL,NMAX):
    P=np.zeros((NMAX))
    ERRO= B-A
    i=0
    while  i < NMAX :
        P[i]=(A+B)/2.0
        if f(P[i])==0 or ERRO<TOL: #Problemas com == 0 (discutir)
            print('iter=',i+1,'Aprox=',P[i])
            print('Erro=',ERRO)
            break
        else:
            if (f(A)*f(P[i])>0):
                A=P[i]
            else:
                B=P[i]
        print('iter=',i+1,'Aprox=',P[i])
        print('Erro=',ERRO)
        ERRO=(B-A)/2 #=(B-P[i])/2
        i+=1
    return P    

### noutro ficheiro
## import bissect
from math import cos

def f(x):
    return cos(x)-x

bissect(f,0.0,1.0,1.e-7,25)


    


Once deleted, variables cannot be recovered. Proceed (y/[n])? y
iter= 1 Aprox= 0.5
Erro= 1.0
iter= 2 Aprox= 0.75
Erro= 0.25
iter= 3 Aprox= 0.625
Erro= 0.125
iter= 4 Aprox= 0.6875
Erro= 0.0625
iter= 5 Aprox= 0.71875
Erro= 0.03125
iter= 6 Aprox= 0.734375
Erro= 0.015625
iter= 7 Aprox= 0.7421875
Erro= 0.0078125
iter= 8 Aprox= 0.73828125
Erro= 0.00390625
iter= 9 Aprox= 0.740234375
Erro= 0.001953125
iter= 10 Aprox= 0.7392578125
Erro= 0.0009765625
iter= 11 Aprox= 0.73876953125
Erro= 0.00048828125
iter= 12 Aprox= 0.739013671875
Erro= 0.000244140625
iter= 13 Aprox= 0.7391357421875
Erro= 0.0001220703125
iter= 14 Aprox= 0.73907470703125
Erro= 6.103515625e-05
iter= 15 Aprox= 0.739105224609375
Erro= 3.0517578125e-05
iter= 16 Aprox= 0.7390899658203125
Erro= 1.52587890625e-05
iter= 17 Aprox= 0.7390823364257812
Erro= 7.62939453125e-06
iter= 18 Aprox= 0.7390861511230469
Erro= 3.814697265625e-06
iter= 19 Aprox= 0.7390842437744141
Erro= 1.9073486328125e-06
iter= 20 Aprox= 0.7390851974487305
Erro= 9.53674

array([0.5       , 0.75      , 0.625     , 0.6875    , 0.71875   ,
       0.734375  , 0.7421875 , 0.73828125, 0.74023438, 0.73925781,
       0.73876953, 0.73901367, 0.73913574, 0.73907471, 0.73910522,
       0.73908997, 0.73908234, 0.73908615, 0.73908424, 0.7390852 ,
       0.73908472, 0.73908496, 0.73908508, 0.73908514, 0.        ])

### Exercício:

- A equação $f(x)= x^3+4x^2-10=0$ tem uma e uma só raiz  em $[1,2]$. Sabe-se que a solução exacta é dada por $p=1.365230013...$ (9 decimais exactos).
- Utilize 13 iterações do método da bissecção para aproximar o valor de $p$. Guarde o valor das aproximações em $p[i]$. 
- Calcule os majorantes dos erros em todas as iterações calculadas.
- Calcule os erros exactos em todas as iterações.
- Calcule $f(p[i]), (i=1\dots 13)$. Atente nos resultados obtidos e comente. 
